iamclient/iamclient.go

```diff
@@ -0,0 +1,353 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iamclient
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"sync"
+	"time"
+
+	pb "github.com/aoscloud/aos_common/api/iamanager/v3"
+	"github.com/golang/protobuf/ptypes/empty"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/aoscloud/aos_common/aoserrors"
+	"github.com/aoscloud/aos_common/utils/cryptutils"
+	"github.com/aoscloud/aos_iamanager/config"
+)
+
+/***********************************************************************************************************************
+ * Consts
+ **********************************************************************************************************************/
+
+const (
+	connectionTimeout = 10 * time.Minute
+	requestTimeout    = 30 * time.Second
+)
+
+/***********************************************************************************************************************
+ * Types
+ **********************************************************************************************************************/
+
+// CertHandler interface.
+type CertHandler interface {
+	GetCertificate(certType string, issuer []byte, serial string) (certURL, keyURL string, err error)
+}
+
+// Client IAM client instance.
+type Client struct {
+	cryptoContext    *cryptutils.CryptoContext
+	certHandler      CertHandler
+	provisioningMode bool
+	dialOptions      []grpc.DialOption
+	remoteIAMs       map[string]*remoteIAM
+	ctx              context.Context // nolint: containedctx // used to keep common context
+	cancelFunc       context.CancelFunc
+}
+
+type remoteIAM struct {
+	sync.Mutex
+
+	connection      *grpc.ClientConn
+	connectionTimer *time.Timer
+	cfg             config.RemoteIAM
+}
+
+/***********************************************************************************************************************
+ * Vars
+ **********************************************************************************************************************/
+
+var errNodeNotFound = errors.New("node not found")
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+// New creates new IAM client instance.
+func New(
+	cfg *config.Config, cryptoContext *cryptutils.CryptoContext, certHandler CertHandler, provisioningMode bool,
+) (client *Client, err error) {
+	client = &Client{
+		cryptoContext:    cryptoContext,
+		certHandler:      certHandler,
+		provisioningMode: provisioningMode,
+		remoteIAMs:       make(map[string]*remoteIAM),
+	}
+
+	defer func() {
+		if err != nil {
+			client.Close()
+		}
+	}()
+
+	if provisioningMode {
```

Comment:
maybe rename it to provisionedMode? Because in provisioning mode we have an insecure connection